fix: beagle helpers page and add beagle grpc tutorials (#611)
ZupIT/beagle-docs: content/en/resources/gRPC/beagle-grpc-web.md

---
title: "Beagle gRPC for Web"
weight: 1
description: >-

>   In this section you will find all the info related to the client and the CLI of gRPC for Beagle Web (Angular and React).

Reviewer: ```suggestion
  In this section, you will find information about the client and the CLI of gRPC for Beagle Web (Angular and React).
```

---

## What is gRPC?

---

It is an open source framework to perform RPC (Remote call procedure) calls, through the HTTP2 protocol in a two-way streaming format, using Protobufs as interfaces between client and server.

> Protobufs, or Protocol Buffers, are a structured data serialization mechanism created by Google, which aim to keep contracts always valid and consistent regarding the implementations that consume a gRPC service.
> 
> Proto files, in their content, are very similar to any interface we are used to on the daily basis, but with the .proto extension, where we define which objects and methods are likely to be used in the service's communication.

Reviewer: ```suggestion
- **Protobufs or Protocol Buffers**: They are structured data serialization mechanisms created by Google, which aim to keep contracts always valid and consistent, regarding the implementations that consume a gRPC service.

- **Proto files**: They are very similar to any interface we are used to on the daily basis, in their content. However, they have the .proto extension, where you can define which objects and methods are likely to be used in the service's communication.
```

[Learn more...](https://grpc.io/)

## gRPC for Beagle Web

---

> gRPC is a technology that makes use of the **HTTP 2** protocol, however, this technology is not compatible with the current Web structure, so the structure created uses a Client and a CLI, where the CLI is responsible for handling gRPC connections through a Proxy, written with Go, and the Client is responsible for handling all requests made by Beagle to the Proxy.

Reviewer: ```suggestion
gRPC is a technology that uses the **HTTP 2** protocol, however, this technology is not compatible with the current Web structure. So, the new structure created uses a Client and a CLI: 
- The CLI is responsible for handling gRPC connections through a Proxy written with Go
- The Client is responsible for handling all requests made by Beagle to the Proxy.
```

## Requirements

---

- [Git - CLI](https://git-scm.com/): Used internally as a proxy dependency.
- [Go - CLI](https://golang.org/): Language which the proxy is written.
- [Dep - CLI](https://github.com/golang/dep): Go's dependency manager, where it checks that everything is installed and working as expected.
- [Protoc - CLI](https://github.com/protocolbuffers/protobuf): Used by protobufs and necessary for code generation through a `.proto` file.

## Installation

---

> To install the library, install it globally, for a better experience with the CLI, and then install it in your project using the following commands:

Reviewer: ```suggestion
Install the library globally. For a better experience with the CLI, install it in your project using the following commands:
```

```bash
npm install -g @zup-it/beagle-web-grpc
npm install --save-dev @zup-it/beagle-web-grpc
```

or if you use yarn:

```bash
yarn global add @zup-it/beagle-web-grpc
yarn add --dev @zup-it/beagle-web-grpc
```

## Getting Started

---

> 1. After installing the library, in the project and globally, create the configuration file in the project's root folder, using: `beagle-web-grpc init` and then configure the attributes according to your needs.
> 2. Then import the gRPC Client:

Reviewer: ```suggestion
To start using, follow the next steps: 
**Step 1.** After installing the library in the project and globally, create the configuration file in the project's root folder, using: **`beagle-web-grpc init`** and then configure the attributes according to your needs.
**Step 2.**  Import the gRPC Client, check out below: 
```

   - **React:** ./src/beagle/beagle-service.ts

     ```typescript
     import { createBeagleUIService } from "@zup-it/beagle-react";
     import { usingBeagleGrpcClient } from "@zup-it/beagle-web-grpc";

     export default createBeagleUIService({
       baseUrl: "grpc://",
       components: {},
       fetchData: usingBeagleGrpcClient({
         proxyAddress: "http://localhost:8081",
       }),
     });
     ```

   - **Angular:** ./src/app/beagle.module.ts

     ```typescript
     import { BeagleModule } from "@zup-it/beagle-angular";
     import { usingBeagleGrpcClient } from "@zup-it/beagle-web-grpc";

     @BeagleModule({
       baseUrl: "grpc://",
       module: {
         path: "./beagle-components.module",
         name: "BeagleComponentsModule",
       },
       components: {},
       fetchData: usingBeagleGrpcClient({
         proxyAddress: "http://localhost:8081",
       }),
     })
     export class Beagle {}
     ```

> 3. Start your proxy running the following command on the root of the project: `beagle-web-grpc start-proxy` or `beagle-web-grpc spx` or `beagle-web-grpc start-proxy --mode your-mode`.
> 4. Use the Beagle Web as usually.

Reviewer: ```suggestion
**Step 3.** Start your proxy running the following command on the root of the project: **`beagle-web-grpc start-proxy`** or **`beagle-web-grpc spx`** or **`beagle-web-grpc start-proxy --mode your-mode`.**
**Step 4.** Use the Beagle Web as usually.
```

> **Observations**
> Make sure that your BFF Application is running.

## CLI

---

> As gRPC is not compatible with the current structure of the Web, to support the technology, it is necessary to use a Proxy to establish the communication between the REST application and the gRPC Backend application, so we created a CLI (command-line interface), **beagle-web-grpc**, to facilitate the use of the proxy and provide all the necessary settings.

Reviewer: ```suggestion
gRPC is not compatible with the current structure of the Web, to support the technology, you have to use a Proxy to establish the communication between the REST application and the gRPC Backend application. 
Beagle team created a CLI (command-line interface), **beagle-web-grpc**, to make it easier to use the proxy and to provide all the necessary settings.
```

| -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `mode`               | Defines the mode name for the settings to be used when running the gRPC Proxy                                                                               |
| `grpcBackendAddress` | Defines the address of the backend server that communicates using gRPC. Remembering that all requests that match this route will be redirected to the Proxy |
| `tlsCertificatePath` | Set path to TLS Certificate location                                                                                                                        |
| `tlsKeyPath`         | Set path to TLS Key location                                                                                                                                |
| `runProxyOnPort`     | Defines the Port on which the gRPC Proxy will run                                                                                                           |

## Client

---

> The gRPC Client for Beagle Web is what redirects grpc requests to the proxy, the client is used through `BeagleService` in React and the `Beagle` module in Angular, using the `fetchData` property, which Beagle allows you to create custom HTTP requests, so the use goes like this:

Reviewer: ```suggestion
The gRPC Client for Beagle Web is what redirects gRPC requests to the proxy. The client is used through `BeagleService` in React and `Beagle`'s module in Angular. When using the `fetchData` property, Beagle allows you to create custom HTTP requests, so it goes like this:
```
